### PR TITLE
Baseline feature

### LINKED ISSUE #9: Baseline
Just like PHPStan we need a baseline functionality.

### PR DESCRIPTION
**Pull request by raffaelecarelle**
Hello @ruudk,

I'worked on baseline generation.

I hopefull it is good enough for you.

Thano you!

Fixes #9

### CODE REVIEW

**Comment by ruudk** [comment]
Thanks @raffaelecarelle for giving this a try.

I'm going to hold with merging this because I want to think about the best strategy to do this.

One solution could be to do it like this PR suggests, to let PHPStan generate the baseline.

Another solution could be, to let TwigStan handle the baseline. Since TwigStan [transforms the errors](https://github.com/twigstan/twigstan/blob/main/src/PHPStan/Analysis/ErrorTransformer.php) (for example, it changes `$variableName` to `variableName` to better match Twig's terminology), it might make more sense that those errors are in the baseline.  

At the moment I'm busy with improving the test suite. I want that to be done first, before continuing here.

**Comment by raffaelecarelle** [comment]
ok @ruudk, I understand this. 

Thank you for your review.

**Comment by raffaelecarelle** [comment]
Hello @ruudk, 

I continued with the baseline work. Twigstan creates its own by no longer using phpstan to generate it. I used some phpstan components to remove errors to ignore from the results. Try taking a look at it. Any advice is welcome.

File: src/Application/PHPStanRunner.php [hunk 1]
```diff
@@ -71,7 +71,7 @@ public function run(
         $process = new Process(array_filter([
             PHP_BINARY,
             $xdebugMode ? '-dzend_extension=xdebug.so' : null,
-            'vendor/bin/phpstan',
+            getcwd() . '/vendor/bin/phpstan',
```
Comment (ruudk) [inline src/Application/PHPStanRunner.php:74]: I wonder why this was done. Because we already pass the working directory on line 86.

File: src/Baseline/Formatter/BaselineErrorFormatterInterface.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace TwigStan\Baseline\Formatter;
+
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use TwigStan\Application\TwigStanAnalysisResult;
+use TwigStan\Application\TwigStanError;
+use PHPStan\Command\Symfony\SymfonyOutput;
+
+interface BaselineErrorFormatterInterface
```
Comment (ruudk) [inline src/Baseline/Formatter/BaselineErrorFormatterInterface.php:11]: Can you remove the Interface suffix?

File: src/Application/AnalyzeCommand.php [hunk 3]
```diff
@@ -4,20 +4,33 @@
...
+use Nette\Neon\Neon;
+use PHPStan\Analyser\Error;
+use PHPStan\Analyser\Ignore\IgnoredErrorHelper;
+use PHPStan\File\CouldNotWriteFileException;
+use PHPStan\File\FileReader;
+use PHPStan\File\FileWriter;
+use PHPStan\File\ParentDirectoryRelativePathHelper;
+use PHPStan\Internal\DirectoryCreator;
+use PHPStan\Internal\DirectoryCreatorException;
+use PHPStan\ShouldNotHappenException;
```
Comment (ruudk) [inline src/Application/AnalyzeCommand.php:16]: We should not depend on PHPStan for this.

File: config/application.neon [hunk 4]
```diff
@@ -45,3 +47,11 @@ services:
 	- TwigStan\Twig\TokenParser\DumpTypeTokenParser
 	- TwigStan\Twig\TwigFactory(environmentLoader: %twigstan.environmentLoader%)
 	- TwigStan\Twig\TwigFileNormalizer
+	- PHPStan\File\FileHelper(
+		workingDirectory: %currentWorkingDirectory%
+	)
```
Comment (ruudk) [inline config/application.neon:52]: Maybe we don't need this. In most places it's handled with `Path::makeRelative`

File: src/Application/AnalyzeCommand.php [hunk 5]
```diff
@@ -31,26 +44,27 @@
...
      * @param list<string> $excludes
      */
     public function __construct(
-        private TwigCompiler $twigCompiler,
-        private TwigFlattener $twigFlattener,
-        private TwigScopeInjector $twigScopeInjector,
-        private DependencyFinder $dependencyFinder,
-        private DependencySorter $dependencySorter,
+        private TwigCompiler       $twigCompiler,
+        private TwigFlattener      $twigFlattener,
```
Comment (ruudk) [inline src/Application/AnalyzeCommand.php:55]: Could you rebase with `main`. These spaces shouldn't be there.

**Review by ruudk — COMMENTED** [review]
Thanks for giving this a try 😊 

It currently directly uses PHPStan code, it should not. 

TwigStan invokes PHPStan via CLI. There is no direct dependency on the code parts. Except for the collectors, those run inside PHPStan.

File: config/application.neon (hunk 4 above)
Comment (raffaelecarelle) [reply config/application.neon]: done

File: src/Application/AnalyzeCommand.php (hunk 3 above)
Comment (raffaelecarelle) [reply src/Application/AnalyzeCommand.php]: Removed all phpstan dependencies

File: src/Baseline/Formatter/BaselineErrorFormatterInterface.php (hunk 2 above)
Comment (raffaelecarelle) [reply src/Baseline/Formatter/BaselineErrorFormatterInterface.php]: done

File: src/Application/PHPStanRunner.php (hunk 1 above)
Comment (raffaelecarelle) [reply src/Application/PHPStanRunner.php]: I don't know why, with my GenerationBaselineTest.php I need this, otherwise something goes wrong (Could not open input file: vendor/bin/phpstan)

File: src/Application/AnalyzeCommand.php [hunk 6]
```diff
@@ -58,6 +61,8 @@ protected function configure(): void
         $this->addArgument('paths', InputArgument::IS_ARRAY | InputArgument::OPTIONAL, 'Directories or files to analyze');
         $this->addOption('debug', null, InputOption::VALUE_NONE, 'Enable debug mode');
         $this->addOption('xdebug', null, InputOption::VALUE_NONE, 'Enable xdebug mode');
+        $this->addOption('generate-baseline', null, InputOption::VALUE_NONE, 'Path to a file where the baseline should be saved');
+        $this->addOption('allow-empty-baseline', null, InputOption::VALUE_OPTIONAL, 'Allow to generate PHPStan empty baseline', false);
```
Comment (ruudk) [inline src/Application/AnalyzeCommand.php:65]: Not sure if we need this option to be honest. If you want a baseline, and there are no errors, then you get an empty baseline.

File: src/Application/AnalyzeCommand.php [hunk 7]
```diff
@@ -29,7 +31,7 @@
 use TwigStan\Twig\TwigFileNormalizer;
 
 #[AsCommand(name: 'analyze', aliases: ['analyse'])]
-final class AnalyzeCommand extends Command
+class AnalyzeCommand extends Command
```
Comment (ruudk) [inline src/Application/AnalyzeCommand.php:34]: Why this change

File: src/Application/AnalyzeCommand.php [hunk 8]
```diff
@@ -444,4 +464,47 @@ private function getFinder(array $paths): array
...
+
+        $baselineFormatted = $baselineErrorFormatter->format($result, $existingBaselineContent);
+
+        try {
+            $this->filesystem->mkdir($baselineFileDirectory, 0644);
+        } catch (IOException $e) {
+            $output->writeln($e->getMessage());
+        }
+
+        $success = @file_put_contents($generateBaselineFile, $baselineFormatted);
```
Comment (ruudk) [inline src/Application/AnalyzeCommand.php:497]: Can we use Filesystem::dumpFile

File: src/Application/AnalyzeCommand.php [hunk 9]
```diff
@@ -444,4 +464,47 @@ private function getFinder(array $paths): array
...
+        $extension = pathinfo($generateBaselineFile, PATHINFO_EXTENSION);
+
+        if ($extension !== 'neon') {
+            throw new \RuntimeException("$extension extension is not supported yet.");
+        }
+
+        $baselineFileDirectory = dirname($generateBaselineFile);
+
+        $baselineErrorFormatter = new BaselineNeonErrorFormatter($this->currentWorkingDirectory);
+        $existingBaselineContent = (string) @file_get_contents($generateBaselineFile);
```
Comment (ruudk) [inline src/Application/AnalyzeCommand.php:487]: Filesystem::readFile

File: src/Application/AnalyzeCommand.php [hunk 10]
```diff
@@ -444,4 +464,47 @@ private function getFinder(array $paths): array
...
+    {
+        if ($generateBaselineFile === false) {
+            return;
+        }
+
+        if ($generateBaselineFile === true) {
+            $generateBaselineFile = 'twigstan-baseline.neon';
+        }
+
+        $extension = pathinfo($generateBaselineFile, PATHINFO_EXTENSION);
```
Comment (ruudk) [inline src/Application/AnalyzeCommand.php:478]: Can we use `\Symfony\Component\Filesystem\Path` for this?

File: src/Application/AnalyzeCommand.php (hunk 6 above)
Comment (raffaelecarelle) [reply src/Application/AnalyzeCommand.php]: you're right. I had removed allow-empty-baseline.

File: src/Application/AnalyzeCommand.php (hunk 7 above)
Comment (raffaelecarelle) [reply src/Application/AnalyzeCommand.php]: Sorry about that. It was a try but now I can restore as before

File: src/Application/AnalyzeCommand.php (hunk 8 above)
Comment (raffaelecarelle) [reply src/Application/AnalyzeCommand.php]: yes, of course

File: src/Application/AnalyzeCommand.php (hunk 9 above)
Comment (raffaelecarelle) [reply src/Application/AnalyzeCommand.php]: yes, of course. done

File: src/Application/AnalyzeCommand.php (hunk 10 above)
Comment (raffaelecarelle) [reply src/Application/AnalyzeCommand.php]: yes, of course. done

File: src/Application/AnalyzeCommand.php [hunk 11]
```diff
@@ -4,7 +4,6 @@
 
 namespace TwigStan\Application;
 
-use InvalidArgumentException;
```
Comment (VincentLanglet) [inline src/Application/AnalyzeCommand.php:7]: @ruudk You should decide if such class should be used or not and let php-cs-fixer doing the job
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/rules/import/global_namespace_import.rst

Comment (ruudk) [reply src/Application/AnalyzeCommand.php:7]: @raffaelecarelle Could you create another PR that configures `global_namespace_import` like this:
```
['import_classes' => true, 'import_constants' => false, 'import_functions' => false]
```
?

We only want to import classes. Constants and functions not.

Comment (VincentLanglet) [reply src/Application/AnalyzeCommand.php:7]: Done in https://github.com/twigstan/twigstan/pull/28

File: src/Application/AnalyzeCommand.php [hunk 12]
```diff
@@ -383,7 +388,15 @@ function ($renderPoints) {
...
         }
 
-        return $result;
+        if ($generateBaselineFile) {
+            $this->generateBaseline($result, $output, $generateBaselineFile);
+        }
+
+        $ignoredErrorHelperResult = $this->ignoredErrorHelper->initialize();
+
+        $ignoredErrorHelperProcessedResult = $ignoredErrorHelperResult->process($result->errors, true, $twigFileNames, false);
```
Comment (ruudk) [inline src/Application/AnalyzeCommand.php:397]: I think it would make more sense to merge these 2 things into a single IgnoredErrorProcessor class.

File: src/Application/Ignore/IgnoredErrorHelperResult.php [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+declare(strict_types=1);
+
+namespace TwigStan\Application\Ignore;
+
+use Symfony\Component\Filesystem\Path;
+use TwigStan\Application\TwigStanError;
+use TwigStan\Twig\SourceLocation;
+
+final class IgnoredErrorHelperResult
```
Comment (ruudk) [inline src/Application/Ignore/IgnoredErrorHelperResult.php:11]: I feel this class can be merged together with IgnoredErrorHelper.

And then call it `IgnoredErrorProcessor`.

File: src/Baseline/Formatter/BaselineErrorFormattable.php [hunk 14]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace TwigStan\Baseline\Formatter;
+
+use TwigStan\Application\TwigStanAnalysisResult;
+
+interface BaselineErrorFormattable
```
Comment (ruudk) [inline src/Baseline/Formatter/BaselineErrorFormattable.php:7]: I would call it Formatter:
```suggestion
interface BaselineErrorFormatter
```

But maybe this is better:
```suggestion
interface BaselineFormatter
```

File: src/Baseline/Formatter/BaselineNeonErrorFormatter.php [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace TwigStan\Baseline\Formatter;
+
+use Nette\DI\Helpers;
+use Nette\Neon\Neon;
+use Symfony\Component\Filesystem\Path;
+use TwigStan\Application\TwigStanAnalysisResult;
+use TwigStan\Application\TwigStanError;
+
+class BaselineNeonErrorFormatter implements BaselineErrorFormattable
```
Comment (ruudk) [inline src/Baseline/Formatter/BaselineNeonErrorFormatter.php:11]:
```suggestion
class NeonBaselineFormatter implements BaselineFormatter
```

File: src/Application/Ignore/IgnoredErrorHelperResult.php [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+{
+    /**
+     * @param list<string> $errors
+     * @param array<array<mixed>> $otherIgnoreErrors
+     * @param array<string, array<array<mixed>>> $ignoreErrorsByFile
+     * @param (string|mixed[])[] $ignoreErrors
```
Comment (ruudk) [inline src/Application/Ignore/IgnoredErrorHelperResult.php:17]: I'm no fan of these untyped arrays.

I was wondering. Why don't we make IgnoredError a value object, that is the representation of the Ignored Error.

Then this becomes:
```suggestion
     * @param list<IgnoredError> $ignoreErrors
```

File: src/Baseline/Formatter/BaselineNeonErrorFormatter.php [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+        if ($existingBaselineContent === '') {
+            return substr($neon, 0, -1);
+        }
+
+        $existingBaselineContentEndOfFileNewlinesMatches = preg_match("~(\n)+$~", $existingBaselineContent, $matches);
+        $existingBaselineContentEndOfFileNewlines = $existingBaselineContentEndOfFileNewlinesMatches !== false
+            ? $matches[0] ?? ''
+            : '';
+
+        return substr($neon, 0, -2) . $existingBaselineContentEndOfFileNewlines;
```
Comment (ruudk) [inline src/Baseline/Formatter/BaselineNeonErrorFormatter.php:83]: What is this all about?

File: src/Baseline/Formatter/BaselineNeonErrorFormatter.php [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    public function format(
+        TwigStanAnalysisResult $analysisResult,
+        string $existingBaselineContent,
+    ): string {
+        if (count($analysisResult->errors) === 0) {
+            return $this->getNeon([], $existingBaselineContent);
+        }
+
+        $fileErrors = [];
+        /** @var TwigStanError $fileSpecificError */
```
Comment (ruudk) [inline src/Baseline/Formatter/BaselineNeonErrorFormatter.php:26]: Should not be needed.

File: src/Application/ContainerFactory.php [hunk 19]
```diff
@@ -51,6 +51,18 @@ public function create(string $tempDirectory): Container
...
+            foreach ($configuration['includes'] as $include) {
+                if(pathinfo($include, PATHINFO_EXTENSION) === 'neon') {
+                    $content = Neon::decodeFile(Path::makeAbsolute($include, Path::getDirectory($this->configurationFile)));
+                    $ignoreErrors = $content['parameters']['twigstan']['ignoreErrors'];
+                    $configuration['parameters']['twigstan']['ignoreErrors'][] = $ignoreErrors;
+                }
+            }
+
+            $configuration['parameters']['twigstan']['ignoreErrors'] = array_merge(...$configuration['parameters']['twigstan']['ignoreErrors']);
+        }
```
Comment (ruudk) [inline src/Application/ContainerFactory.php:64]: I'm not so sure about this. 

I wonder if this is even needed, given that includes are already supported?

**Review by ruudk — COMMENTED** [review]
Thanks for giving this a shot. I pushed a few changes and added a couple of comments.

It might also be good to rebase with main, and fix the CS after the last changes there.

File: src/Application/Ignore/IgnoredErrorHelperResult.php (hunk 16 above)
Comment (ruudk) [reply src/Application/Ignore/IgnoredErrorHelperResult.php:17]: The benefit is that we can now have type safety on the IgnoredError value object. And move some of the helper methods there too. This might improve maintainability.

File: src/Application/ContainerFactory.php (hunk 19 above)
Comment (raffaelecarelle) [reply src/Application/ContainerFactory.php]: I also tried without that, but ignoreError is always empty

**Comment by raffaelecarelle** [comment]
Hi @ruudk, if you want, take a look :)
